File: cpp/src/VendingMachine.cpp

#include "VendingMachine.h"
#include <sstream>
#include <iomanip>
#include <algorithm>

VendingMachine::VendingMachine() :
        display("INSERT COIN"),
        balance(0L),
        coins(new std::vector<int>(0, 0)),
        returns(new std::vector<int>(0, 0)),
        acceptedCoins(new std::vector<int>(3, 0))
        {
    int temp[] = {5, 10, 25};
    std::copy ( temp, temp+3, acceptedCoins->begin() );
}

static std::string getFormattedNumberAsString(double number, int precision)
{
    std::stringstream stream;
    stream << std::fixed << std::setprecision(precision) << number;
    return stream.str();
}

void VendingMachine::insertCoin(int value) {
    if (std::find(acceptedCoins->begin(), acceptedCoins->end(), value) != acceptedCoins->end()) {
        balance += value;
        coins->push_back(value);
        double dollars = (double)balance/100.0;
        display = "$" + getFormattedNumberAsString(dollars, 2);

    } else {
        returns->push_back(value);
    }
}

VendingMachine::~VendingMachine() {
    delete coins;
    delete returns;
    delete acceptedCoins;
}

File: csharp/VendingMachine-Approval-Kata/VendingMachine.cs
﻿using System.Globalization;

namespace VendingMachine_Approval_Kata;

public class VendingMachine
{
    public List<int> Coins { get; } = new();
    public string Display { get; private set; }
    public int Balance { get; private set; }

    private readonly CultureInfo _en_Us_Culture;
    
    public VendingMachine()
    {
        Display = "";
        _en_Us_Culture = CultureInfo.CreateSpecificCulture("en-US");

        DisplayBalance();
    }

    private void DisplayBalance()
    {
        Display = Balance != 0 ? FormatAsDollars(Balance) : "INSERT COIN";
    }

    private string FormatAsDollars(int cents)
    {
        return (cents / 100.0).ToString("C", _en_Us_Culture);
    }

    public void InsertCoin(int coin)
    {
        Coins.Add(coin);
        Balance += coin;
        DisplayBalance();
    }
}